### PR TITLE
deprecated attribute

### PR DESCRIPTION
**Pull request by xunilrj**
## Description

This PR is part of https://github.com/FuelLabs/sway/pull/4794. It is the minimum implementation for deprecation. Pretty much only what was needed for deprecating `U256`.

It also minimally implements `#[allow(deprecated)]`.

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by xunilrj** [comment]
Reminder to ping https://github.com/FuelLabs/sway/pull/4779 when this is merged.

**Review by IGI-111 — APPROVED** [review]

File: sway-core/src/language/ty/expression/expression.rs [hunk 1]
```diff
@@ -429,4 +433,81 @@ impl TyExpression {
...
+                            &s.attributes,
+                            &call_path.span(),
+                            handler,
+                            "deprecated struct",
+                            allow_deprecated,
+                        );
+                    }
+                }
+            }
+            _ => {}
```
Comment (ironcev) [inline sway-core/src/language/ty/expression/expression.rs:510]: Perhaps just adding a TODO that we want to cover other deprecated items in the future.

**Review by ironcev — DISMISSED** [review]
Nice minimal implementation of `deprecated`! Just a few minor remarks from my side but otherwise LGTM

**Comment by ironcev** [comment]
Since the `UsingDepricated` warning is brand new providing `Diagnostic` for it from the beginning will save us effort down the road. @xunilrj If you are find with that I can add it.

File: sway-core/src/transform/attribute.rs [hunk 2]
```diff
@@ -136,3 +144,36 @@ impl std::ops::Deref for AttributesMap {
...
+#[derive(Default)]
+pub struct AllowDeprecatedState {
+    allowed: u32,
+}
+impl AllowDeprecatedState {
+    pub(crate) fn enter(&mut self, attributes: AttributesMap) -> AllowDeprecatedEnterToken {
+        if let Some(all_allows) = attributes.get(&AttributeKind::Allow) {
+            for allow in all_allows {
+                for arg in allow.args.iter() {
+                    if arg.name.as_str() == "deprecated" {
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply sway-core/src/transform/attribute.rs:161]: Nice catch. Done.

File: sway-core/src/language/ty/expression/expression.rs (hunk 1 above)
Comment (xunilrj) [reply sway-core/src/language/ty/expression/expression.rs:510]: Done.

**Comment by xunilrj** [comment]
> Since the `UsingDepricated` warning is brand new providing `Diagnostic` for it from the beginning will save us effort down the road. @xunilrj If you are find with that I can add it.

I can do it. Do we have a guide I can follow?
I will merge this PR and create a new one, just to finish this.

**Review by IGI-111 — APPROVED** [review]

**Pull request merged**